### DISCUSSION TITLE
setAddress() hw/sw variation?

### DISCUSSION THREAD

**Discussion by Mike-K8LH** [TomNisbet/TommyPROM#65] [Ideas]
Hope I'm not being "a pain", Tom, but I'm hoping someone can suggest how to measure the time required for this variation of the setAddress() function?   It loads both address shift registers in parallel in a single loop with only a minor modification to the original hardware.  TIA, Mike.

```
  /***********************************************************************
   *  PromAddressDriver.cpp for experimental 74HC164 hardware config'    *
   *                                                                     *
   *    'D1' (PD3) -> address 'hi' shift register 'data' pin             *
   *    'D0' (PD2) -> address 'lo' shift register 'data' pin             *
   *   'CLK' (PC3) -> address both shift register 'clock' pins           *
   ***********************************************************************/

   #include "PromAddressDriver.h"

   #define cbi(sfr, bit) (_SFR_BYTE(sfr) &= ~(1<<bit))  // clear bit
   #define sbi(sfr, bit) (_SFR_BYTE(sfr) |= 1<<bit)     // set bit

  /*  bit/pin masks for experimental hardware configuration              */

   #define ADDR_CLOCK_MASK     0b00001000   // A3/PC3 'clock' bit mask

   void PromAddressDriver::begin()
   { DDRB |= ADDR_CLOCK_MASK;               // make 'clock' output
     PORTB &= ~ADDR_CLOCK_MASK;             // make 'clock' lo
   }                                        //

   void PromAddressDriver::setAddress(uint32_t address)
   { byte ix = 8;                           //
     byte upper = (address >> 16) & 0x07;   // set A16..A18 (PB3..PB5)
     PORTB = (PORTB & ~0x1C) | (upper << 2); //  "
     DDRD |= 0b00001100;                    // make 'D1-D0' outputs
     do                                     // load address registers
     { PORTD &= 0b11110011;                 // clr 'D1' & 'D0' pins
       if (address & 0x8000) sbi(PORTD,3);  // set 'D1' pin (hi 164)
       if (address & 0x0080) sbi(PORTD,2);  // set 'D0' pin (lo 164)
       PORTC |= ADDR_CLOCK_MASK;            // toggle 'clock' (PC3)
       delayMicroseconds(3);                //  "
       PORTC &= ~ADDR_CLOCK_MASK;           //  "
       address <<= 1;                       // shift 'address'
     } while(--ix);                         //
     DDRD &= 0b11110011;                    // make 'D1-D0' inputs
   }                                        //
```
![TommyPROM Mod 1](https://github.com/TomNisbet/TommyPROM/assets/38607118/2b1f77c2-2f33-42a9-90fc-7c202f62c7ef)

**Comment by TomNisbet** [comment]
You could easily measure this with a logic analyzer or a storage scope.  If you look at the diagram on the [Software Data Protection page](https://tomnisbet.github.io/TommyPROM/docs/28C256-notes), the existing code takes about 73us to write a byte, and the address bytes are maybe 25us each, so your design would probably cut about a third of the time out of the byte write cycle.

Note that the existing code only writes the high address register when the value changes, so it is already performing similarly to your code for chips like the 28C256 that are writing to sequential addresses.  You should see a performance gain for chips that use a command sequence, like the ST39SF series, because the high address byte changes with every byte programmed.

Even without any analysis tools, you should be able to get a good sense of the change by just programming an entire chip and timing it with a stopwatch.  I'd be interested to see the old code vs new code numbers using the Fill command, because that wouldn't have any data communication delays.
